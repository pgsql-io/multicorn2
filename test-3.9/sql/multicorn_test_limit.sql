SET client_min_messages=NOTICE;
CREATE EXTENSION multicorn;
CREATE server multicorn_srv foreign data wrapper multicorn options (
    wrapper 'multicorn.testfdw.TestForeignDataWrapper'
);
CREATE user mapping FOR current_user server multicorn_srv options (usermapping 'test');

CREATE foreign table testmulticorn (
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'true',
    cansort 'true'
);

CREATE foreign table testmulticorn_nolimit(
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'false',
    cansort 'true'
);

CREATE foreign table testmulticorn_nosort(
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'true',
    cansort 'false'
);

-- Verify limit and offset are not pushed down when FWD says "no" (but still returns the 1 correct row)
EXPLAIN SELECT * FROM testmulticorn_nolimit LIMIT 1 OFFSET 1;
SELECT * FROM testmulticorn_nolimit LIMIT 1 OFFSET 1;

-- Verify limit and offset are pushed down when FWD says "yes" (returns just 1 row with proper cost)
EXPLAIN SELECT * FROM testmulticorn LIMIT 1 OFFSET 1;
SELECT * FROM testmulticorn LIMIT 1 OFFSET 1;

-- Verify limit and offset w/ sort not pushed down (still returns the 1 correct row)
EXPLAIN SELECT * FROM testmulticorn_nolimit ORDER BY test1 LIMIT 1 OFFSET 1;
SELECT * FROM testmulticorn_nolimit ORDER BY test1 LIMIT 1 OFFSET 1;

-- Verify limit and offset w/ sort are pushed down (returns just 1 row with proper cost)
EXPLAIN SELECT * FROM testmulticorn ORDER BY test1 LIMIT 1 OFFSET 1;
SELECT * FROM testmulticorn ORDER BY test1 LIMIT 1 OFFSET 1;

-- Verify limit and offset are not pushed when sort is not pushed down
EXPLAIN SELECT * FROM testmulticorn_nosort ORDER BY test1 LIMIT 1 OFFSET 1;
SELECT * FROM testmulticorn_nosort ORDER BY test1 LIMIT 1 OFFSET 1;

DROP USER MAPPING FOR current_user SERVER multicorn_srv;
DROP EXTENSION multicorn cascade;
