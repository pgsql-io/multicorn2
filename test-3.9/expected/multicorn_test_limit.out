SET client_min_messages=NOTICE;
CREATE EXTENSION multicorn;
CREATE server multicorn_srv foreign data wrapper multicorn options (
    wrapper 'multicorn.testfdw.TestForeignDataWrapper'
);
CREATE user mapping FOR current_user server multicorn_srv options (usermapping 'test');
CREATE foreign table testmulticorn (
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'true',
    cansort 'true'
);
CREATE foreign table testmulticorn_nolimit(
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'false',
    cansort 'true'
);
CREATE foreign table testmulticorn_nosort(
    test1 date,
    test2 timestamp
) server multicorn_srv options (
    option1 'option1',
    test_type 'date',
    canlimit 'true',
    cansort 'false'
);
-- Verify limit and offset are not pushed down when FWD says "no" (but still returns the 1 correct row)
EXPLAIN SELECT * FROM testmulticorn_nolimit LIMIT 1 OFFSET 1;
NOTICE:  [('canlimit', 'false'), ('cansort', 'true'), ('option1', 'option1'), ('test_type', 'date'), ('usermapping', 'test')]
NOTICE:  [('test1', 'date'), ('test2', 'timestamp without time zone')]
                                     QUERY PLAN                                     
------------------------------------------------------------------------------------
 Limit  (cost=29.50..49.00 rows=1 width=20)
   ->  Foreign Scan on testmulticorn_nolimit  (cost=10.00..400.00 rows=20 width=20)
(2 rows)

SELECT * FROM testmulticorn_nolimit LIMIT 1 OFFSET 1;
NOTICE:  []
NOTICE:  ['test1', 'test2']
   test1    |          test2           
------------+--------------------------
 02-03-2011 | Tue Feb 01 14:30:25 2011
(1 row)

-- Verify limit and offset are pushed down when FWD says "yes" (returns just 1 row with proper cost)
EXPLAIN SELECT * FROM testmulticorn LIMIT 1 OFFSET 1;
NOTICE:  [('canlimit', 'true'), ('cansort', 'true'), ('option1', 'option1'), ('test_type', 'date'), ('usermapping', 'test')]
NOTICE:  [('test1', 'date'), ('test2', 'timestamp without time zone')]
                             QUERY PLAN                             
--------------------------------------------------------------------
 Foreign Scan on testmulticorn  (cost=10.00..20.00 rows=1 width=20)
(1 row)

SELECT * FROM testmulticorn LIMIT 1 OFFSET 1;
NOTICE:  []
NOTICE:  ['test1', 'test2']
   test1    |          test2           
------------+--------------------------
 02-03-2011 | Tue Feb 01 14:30:25 2011
(1 row)

-- Verify limit and offset w/ sort not pushed down (still returns the 1 correct row)
EXPLAIN SELECT * FROM testmulticorn_nolimit ORDER BY test1 LIMIT 1 OFFSET 1;
                                     QUERY PLAN                                     
------------------------------------------------------------------------------------
 Limit  (cost=29.50..49.00 rows=1 width=20)
   ->  Foreign Scan on testmulticorn_nolimit  (cost=10.00..400.00 rows=20 width=20)
(2 rows)

SELECT * FROM testmulticorn_nolimit ORDER BY test1 LIMIT 1 OFFSET 1;
NOTICE:  []
NOTICE:  ['test1', 'test2']
NOTICE:  requested sort(s): 
NOTICE:  SortKey(attname='test1', attnum=1, is_reversed=False, nulls_first=False, collate=None)
   test1    |          test2           
------------+--------------------------
 01-01-2011 | Sun Jan 02 14:30:25 2011
(1 row)

-- Verify limit and offset w/ sort are pushed down (returns just 1 row with proper cost)
EXPLAIN SELECT * FROM testmulticorn ORDER BY test1 LIMIT 1 OFFSET 1;
                             QUERY PLAN                             
--------------------------------------------------------------------
 Foreign Scan on testmulticorn  (cost=10.00..20.00 rows=1 width=20)
(1 row)

SELECT * FROM testmulticorn ORDER BY test1 LIMIT 1 OFFSET 1;
NOTICE:  []
NOTICE:  ['test1', 'test2']
NOTICE:  requested sort(s): 
NOTICE:  SortKey(attname='test1', attnum=1, is_reversed=False, nulls_first=False, collate=None)
   test1    |          test2           
------------+--------------------------
 01-01-2011 | Sun Jan 02 14:30:25 2011
(1 row)

-- Verify limit and offset are not pushed when sort is not pushed down
EXPLAIN SELECT * FROM testmulticorn_nosort ORDER BY test1 LIMIT 1 OFFSET 1;
NOTICE:  [('canlimit', 'true'), ('cansort', 'false'), ('option1', 'option1'), ('test_type', 'date'), ('usermapping', 'test')]
NOTICE:  [('test1', 'date'), ('test2', 'timestamp without time zone')]
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Limit  (cost=400.20..400.20 rows=1 width=12)
   ->  Sort  (cost=400.20..400.25 rows=20 width=12)
         Sort Key: test1
         ->  Foreign Scan on testmulticorn_nosort  (cost=10.00..400.00 rows=20 width=20)
(4 rows)

SELECT * FROM testmulticorn_nosort ORDER BY test1 LIMIT 1 OFFSET 1;
NOTICE:  []
NOTICE:  ['test1', 'test2']
   test1    |          test2           
------------+--------------------------
 01-01-2011 | Sun Jan 02 14:30:25 2011
(1 row)

DROP USER MAPPING FOR current_user SERVER multicorn_srv;
DROP EXTENSION multicorn cascade;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to server multicorn_srv
drop cascades to foreign table testmulticorn
drop cascades to foreign table testmulticorn_nolimit
drop cascades to foreign table testmulticorn_nosort
